--- iproute2-3.x/include/linux/fib_rules.h
+++ iproute2-3.x/include/linux/fib_rules.h
@@ -64,6 +64,10 @@ enum {
 	FR_ACT_BLACKHOLE,	/* Drop without notification */
 	FR_ACT_UNREACHABLE,	/* Drop with ENETUNREACH */
 	FR_ACT_PROHIBIT,	/* Drop with EACCES */
+	FR_ACT_RES8,
+	FR_ACT_RES9,
+	FR_ACT_RES10,
+	FR_ACT_FAILED_POLICY,	/* Drop with EPERM */
 	__FR_ACT_MAX,
 };
 
--- iproute2-3.x/include/linux/rtnetlink.h
+++ iproute2-3.x/include/linux/rtnetlink.h
@@ -203,6 +203,7 @@ enum {
 	RTN_THROW,		/* Not in this table		*/
 	RTN_NAT,		/* Translate this address	*/
 	RTN_XRESOLVE,		/* Use external resolver	*/
+	RTN_FAILED_POLICY,	/* Source address failed policy */
 	__RTN_MAX
 };
 
--- iproute2-3.x/ip/rtm_map.c
+++ iproute2-3.x/ip/rtm_map.c
@@ -49,6 +49,8 @@ char *rtnl_rtntype_n2a(int id, char *buf
 		return "nat";
 	case RTN_XRESOLVE:
 		return "xresolve";
+	case RTN_FAILED_POLICY:
+		return "failed_policy";
 	default:
 		snprintf(buf, len, "%d", id);
 		return buf;
@@ -84,6 +86,8 @@ int rtnl_rtntype_a2n(int *id, char *arg)
 		res = RTN_UNICAST;
 	else if (strcmp(arg, "throw") == 0)
 		res = RTN_THROW;
+	else if (strcmp(arg, "failed_policy") == 0)
+		res = RTN_FAILED_POLICY;
 	else {
 		res = strtoul(arg, &end, 0);
 		if (!end || end == arg || *end || res > 255)
